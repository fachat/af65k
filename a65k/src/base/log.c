/****************************************************************************

    generic logging interface
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>
#include <strings.h>
#include <stdarg.h>
#include <ctype.h>

#include "infiles.h"
#include "log.h"


static err_level level = LEV_WARN;

void log_module_init(err_level l) {
	level = l;
}

void log_set_level(err_level l) {
	level = l;
}


void log_term(const char *msg) {
	
	int newline = 0;

	printf(">>>: ");
	char c;
	while ((c = *msg) != 0) {
		if (newline) {
			printf("\n>>>: ");
		}
		if (isprint(c)) {
			putchar(c);
			newline = 0;
		} else
		if (c == 10) {
			newline = 1;
		} else
		if (c == 13) {
			// silently ignore
		} else {
			printf("{%02x}", ((int)c) & 0xff);
			newline = 0;
		}
		msg++;
	}
	printf("\n");
}

void log_errno(const char *msg, ...) {
       va_list args;
       va_start(args, msg);
	char buffer[1024];

	if (index(msg, '%') != NULL) {
		// the msg contains parameter, so we need to fix it
		vsprintf(buffer, msg, args);
		msg = buffer;
	}
        printf(">> %s: errno=%d: %s\n", msg, errno, strerror(errno));
}

void log_x(err_level lev, const char *msg, ...) {

	const char *prefix = ">>>";
	switch (lev) {
	case LEV_TRACE: prefix = "TRC:"; break;
	case LEV_DEBUG: prefix = "DBG:"; break;
	case LEV_INFO: 	prefix = "INF:"; break;
	case LEV_WARN: 	prefix = "WRN:"; break;
	case LEV_ERROR: prefix = "ERR:"; break;
	case LEV_FATAL: prefix = "FTL:"; break;
	}

	if (lev >= level) {
       		va_list args;
       		va_start(args, msg);

       		printf(prefix);
       		vprintf(msg, args);
		printf("\n");
	}
}




