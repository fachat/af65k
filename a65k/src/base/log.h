/****************************************************************************

    generic logging interface
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef LOG_H
#define LOG_H

void log_errno(const char *msg, ...);

void log_error(const char *msg, ...);

#ifdef DEBUG

void log_warn(const char *msg, ...);

void log_info(const char *msg, ...);

void log_debug(const char *msg, ...);

void log_term(const char *msg);

#define	log_rv(rv)	log_error("ERROR RETURN: %d\n", (rv))

#else

#define	log_warn(msg, ...)
#define	log_info(msg, ...)
#define	log_debug(msg, ...)
#define	log_term(msg, ...)
#define	log_rv(rv)

#endif

#endif

