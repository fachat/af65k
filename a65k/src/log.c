/****************************************************************************

    generic logging interface
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <strings.h>
#include <stdarg.h>
#include <ctype.h>

void log_term(const char *msg) {
	
	int newline = 0;

	printf(">>>: ");
	char c;
	while ((c = *msg) != 0) {
		if (newline) {
			printf("\n>>>: ");
		}
		if (isprint(c)) {
			putchar(c);
			newline = 0;
		} else
		if (c == 10) {
			newline = 1;
		} else
		if (c == 13) {
			// silently ignore
		} else {
			printf("{%02x}", ((int)c) & 0xff);
			newline = 0;
		}
		msg++;
	}
	printf("\n");
}

void log_errno(const char *msg, ...) {
       va_list args;
       va_start(args, msg);
	char buffer[1024];

	if (index(msg, '%') != NULL) {
		// the msg contains parameter, so we need to fix it
		vsprintf(buffer, msg, args);
		msg = buffer;
	}
        printf(">> %s: errno=%d: %s\n", msg, errno, strerror(errno));
}

void log_warn(const char *msg, ...) {
       va_list args;
       va_start(args, msg);

       printf("WRN:");
        vprintf(msg, args);
}

void log_error(const char *msg, ...) {
       va_list args;
       va_start(args, msg);

       printf("ERR:");
        vprintf(msg, args);
}

void log_info(const char *msg, ...) {
       va_list args;
       va_start(args, msg);

       printf("INF:");
        vprintf(msg, args);
}

void log_debug(const char *msg, ...) {
       va_list args;
       va_start(args, msg);

       printf("DBG:");
        vprintf(msg, args);
}



