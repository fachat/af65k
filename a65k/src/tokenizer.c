/****************************************************************************

    tokenizing a line
    Copyright (C) 2015 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

// this code simply tokenizes the input string given some simple basic rules.

#include <ctype.h>
#include <stdbool.h>

#include "types.h"
#include "mem.h"
#include "err.h"
#include "tokenizer.h"


static type_t tokenizer_memtype = {
	"tokenizer_t",
	sizeof(tokenizer_t)
};

// initialize a tokenizer 
tokenizer_t *tokenizer_init(const char *line) {
	
	tokenizer_t *tok = mem_alloc(&tokenizer_memtype);

	tok->line = line;
	tok->ptr = 0;
	tok->len = 0;
	tok->type = T_INIT;

	return tok;
}

static bool_t parse_base(tokenizer_t *tok, int ptr, int base) {

	int limit = base - 10;
	const char *line = tok->line;
	signed char c = 0;

	long value = 0;
	while (line[ptr] != 0) {
		if (isdigit(c = tok->line[ptr])) {
			c &= 0x0f;
			if (c > (base-1)) {
				// TODO: continue till end of number, so simple typo still allows detect further errors
				tok->vals.errno = E_TOK_DIGITRANGE;
				tok->type = T_ERROR;
				return false;
			}
			value = value * base + c;
		} else 
		if (base > 10) {
			c = tolower(c) - 'a';
		
			// c is char, which is unsigned, so no need to compare for <0
			if (c < limit) {
				value = value * base + c;
			} else {
				break;
			}
		} else {
			break;
		}
		ptr++;
	}
	tok->vals.literal.value = value;
	tok->len = ptr - tok->ptr;

	if (tok->len == 0) {
		// syntax error
		tok->vals.errno = E_TOK_EMPTY;
		tok->type = T_ERROR;
		return false;
	}
	return true;
}

static inline bool_t parse_decimal(tokenizer_t *tok, int ptr) {

	tok->type = T_LITERAL;
	tok->vals.literal.type = LIT_DECIMAL;
	return parse_base(tok, ptr, 10);
}

static inline bool_t parse_octal(tokenizer_t *tok, int ptr) {

	tok->type = T_LITERAL;
	tok->vals.literal.type = LIT_OCTAL;
	return parse_base(tok, ptr, 8);
}

static inline bool_t parse_binary(tokenizer_t *tok, int ptr) {

	tok->type = T_LITERAL;
	tok->vals.literal.type = LIT_BINARY;
	return parse_base(tok, ptr, 2);
}

static inline bool_t parse_hex(tokenizer_t *tok, int ptr) {

	tok->type = T_LITERAL;
	tok->vals.literal.type = LIT_HEX;
	return parse_base(tok, ptr, 16);
}

static inline bool_t parse_string(tokenizer_t *tok, int ptr) {

	const char *line = tok->line;

	char delim = line[ptr];
	tok->type = T_STRING;
	tok->vals.string.type = delim;

	ptr++;
	tok->vals.string.ptr = ptr;
	tok->vals.string.len = 0;

	while (isPrint(line[ptr])) {
		if (line[ptr] == delim) {
			break;
		}
		ptr++;
	}
	tok->vals.string.len = ptr - tok->vals.string.ptr;
	
	if (line[ptr] == delim) {
		ptr++;
	} else 
	if (line[ptr] != 0) {
		// non-printable char ends string
		tok->vals.errno = E_TOK_NONPRINT;
		tok->type = T_ERROR;
		return false;
	}

	tok->len = ptr - tok->ptr;

	if (tok->vals.string.len == 0) {
		// empty string
		tok->vals.errno = E_TOK_EMPTY;
		tok->type = T_ERROR;
		return false;
	}
	return true;
}

static inline bool_t parse_name(tokenizer_t *tok, int ptr) {

	const char *line = tok->line;
	tok->type = T_NAME;

	while (isalnum(line[ptr]) || (line[ptr] == '_')) {
		ptr++;
	}

	tok->len = ptr - tok->ptr;

	return true;
}

static inline bool_t parse_token(tokenizer_t *tok, int ptr, int can_have_operator, int allow_index) {

	const char *line = tok->line;

	char c = line[ptr];

	// default outcome
	tok->vals.op = c;
	tok->type = T_TOKEN;
	tok->len = 1;

	switch(c) {
	case '@':
	case '`':
	case '\'':
	case '#':
	case ':':
	case '.':
	case ';':
		return true;
	case '(':
	case ')':
		tok->type = T_BRACKET;
		return true;
	case ',':
		// TODO: check OP_*IND
		// using allow_index
		return true;
	case '%':
		if (line[ptr+1] == '=') {
			tok->vals.op = OP_ASSIGNMOD;
			tok->len++;
			return true;
		}
		return true;
	case '-':
		if (line[ptr+1] == '=') {
			tok->vals.op = OP_ASSIGNSUB;
			tok->len++;
			return true;
		}
		return true;
	case '+':
		if (line[ptr+1] == '=') {
			tok->vals.op = OP_ASSIGNADD;
			tok->len++;
			return true;
		}
		return true;
	case '*':
		if (line[ptr+1] == '=') {
			tok->vals.op = OP_ASSIGNMULT;
			tok->len++;
			return true;
		}
		return true;
	case '/':
		switch (line[ptr+1]) {
		case '=':
			tok->vals.op = OP_ASSIGNDIV;
			tok->len++;
			return true;
		case '/':
			tok->vals.op = OP_DOUBLESLASH;
			tok->len++;
			return true;
		}
		return true;
	case ']':
		tok->type = T_BRACKET;
		if (line[ptr+1] == ']') {
			tok->vals.op = OP_BBCLOSE;
			tok->len++;
			return true;
		}
		// OP_BOPEN
		return true;
	case '[':
		tok->type = T_BRACKET;
		if (line[ptr+1] == '[') {
			tok->vals.op = OP_BBOPEN;
			tok->len++;
			return true;
		}
		// OP_BOPEN
		return true;
	case '!':
		if (can_have_operator) {
			switch (line[ptr+1]) {
			case '=':
				tok->vals.op = OP_NOTEQUAL;
				tok->len++;
				return true;
			default:
				// default
				// OP_EXCL
				return true;
			}
		} else {
			// OP_EXCL
			return true;
		}
	case '>':
		if (can_have_operator) {
			// check ">>", ">=", "><"
			switch (line[ptr+1]) {
			case '>':
				if (line[ptr+2] == '=') {
					tok->vals.op = OP_ASSIGNSHIFTRIGHT;
					tok->len += 2;
					return true;
				}
				tok->vals.op = OP_SHIFTRIGHT;
				tok->len++;
				return true;
			case '=':
				tok->vals.op = OP_LARGEROREQUAL;
				tok->len++;
				return true;
			case '<':
				tok->vals.op = OP_NOTEQUAL;
				tok->len++;
				return true;
			default:
				// default
				return true;
			}
		} else {
			// ">" for high byte modifier or larger than
			return true;
		}
	case '<':
		if (can_have_operator) {
			// check "<<", "<=", "<>"
			switch (line[ptr+1]) {
			case '<':
				if (line[ptr+2] == '=') {
					tok->vals.op = OP_ASSIGNSHIFTLEFT;
					tok->len += 2;
					return true;
				}
				tok->vals.op = OP_SHIFTLEFT;
				tok->len++;
				return true;
			case '=':
				tok->vals.op = OP_LESSOREQUAL;
				tok->len++;
				return true;
			case '>':
				tok->vals.op = OP_NOTEQUAL;
				tok->len++;
				return true;
			default:
				// default
				return true;
			}
		} else {
			// "<" for high byte modifier
			return true;
		}
	case '=':
		if (can_have_operator) {
			// check "==", "=>", "=<"
			switch (line[ptr+1]) {
			case '=':
				tok->vals.op = OP_EQUAL;
				tok->len++;
				return true;
			default:
				// default, assign
				return true;
			}
		} else {
			// '=' assign
			return true;
		}
	case '&':
		// check "&", "&&", "&="
		switch (line[ptr+1]) {
		case '&':
			tok->vals.op = OP_LOGICAND;
			tok->len++;
			return true;
		case '=':
			tok->vals.op = OP_ASSIGNBITAND;
			tok->len++;
			return true;
		} 
		return true;
	case '|':
		// check "|", "||", "|="
		switch (line[ptr+1]) {
		case '|':
			tok->vals.op = OP_LOGICOR;
			tok->len++;
			return true;
		case '=':
			tok->vals.op = OP_ASSIGNBITOR;
			tok->len++;
			return true;
		} 
		return true;
	case '^':
		// check "^", "^^", "^="
		switch (line[ptr+1]) {
		case '^':
			tok->vals.op = OP_LOGICXOR;
			tok->len++;
			return true;
		case '=':
			tok->vals.op = OP_ASSIGNBITXOR;
			tok->len++;
			return true;
		} 
		return true;
	default:
		tok->vals.errno = E_TOK_UNKNOWN;
		tok->type = T_ERROR;
		return false;
	}

	return true;
}

// set to next token; return true when there is a valid token
bool_t tokenizer_next(tokenizer_t *tok, int allow_index) {

	// move behind last token
	int ptr = tok->ptr + tok->len;

	const char *line = tok->line;

	// attempt to not have to store an extra operand/operator flag in the tokenizer	
	// or even have to pass it to this function as parameter...
	bool_t can_have_operator = (tok->type == T_LITERAL) 
			|| (tok->type == T_TOKEN && line[ptr] == ')');

	// skip whitespace
	while (line[ptr] != 0 && isspace(line[ptr])) {
		ptr++;
	}

	tok->ptr = ptr;

	//printf("ptr=%d, c=%d\n", ptr, line[ptr]);

	// are we done with the line yet?
	if (line[ptr] == 0 || tok->type == T_END || tok->type == T_ERROR) {
		tok->len = 0;
		tok->type = T_END;
		return false;
	}

	char c = line[ptr];

	if (isdigit(c)) {
		// handle octal (starting with '0'), hex (starting with '0x'), dec
		if (c == '0') {
			if (line[ptr+1] == 'x' || line[ptr+1] == 'X') {
				// hex
				ptr += 2;
				tok->ptr = ptr;
				return parse_hex(tok, ptr);
			} else {
				// octal
				return parse_octal(tok, ptr);
			}
		} else {
			// dec
			return parse_decimal(tok, ptr);
		}
	} else
	if ((!can_have_operator) && (c == '%')) {
		// binary
		ptr++;
		tok->ptr = ptr;
		return parse_binary(tok, ptr);
	} else
	if ((!can_have_operator) && (c == '$')) {
		// hex
		ptr++;
		tok->ptr = ptr;
		return parse_hex(tok, ptr);
	} else
	if ((!can_have_operator) && (c == '&')) {
		// octal
		ptr++;
		tok->ptr = ptr;
		return parse_octal(tok, ptr);
	} else
	if (is_string_delim(c)) {
		// string literal
		// c must be of one of quotetype_t enum!
		return parse_string(tok, ptr);
	} else
	if (isalpha(c) || c == '_') {
		// name
		return parse_name(tok, ptr);
	} else
	if (ispunct(c)) {
		// any other token
		return parse_token(tok, ptr, can_have_operator, allow_index);
	} else {
		// non-printable - error
		tok->vals.errno = E_TOK_UNKNOWN;
		tok->type = T_ERROR;
		return false;
	}
	
}


void tokenizer_free(tokenizer_t *tok) {
	
	tok->line = NULL;

	mem_free(tok);
}



