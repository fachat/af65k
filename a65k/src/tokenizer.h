/****************************************************************************

    tokenizing a line
    Copyright (C) 2015 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef TOKENIZER_H
#define TOKENIZER_H

#include <stdio.h>

#include "types.h"

typedef enum {
	T_INIT			=  0,	// initial state
	T_NAME			=  1,	// a name, like label1 do_this
	T_TOKEN			=  2,	// a token, like ( ) . ++ == <> & 
	T_STRING_LITERAL	= 10,	// a quoted string, like 'abc' "cde"
	T_DECIMAL_LITERAL	= 20,	// a decimal number, like 1 2 3
	T_OCTAL_LITERAL		= 28,	// an octal number, like &127 0127
	T_BINARY_LITERAL	= 22,	// a binary number, like %110101
	T_HEX_LITERAL		= 26,	// a hex number, like $12dA 0x12Ad
	T_ERROR			=254,	// an error has happened
	T_END			=255,	// end of string reached
} tok_t;

typedef enum {
	OP_NONE			= 0,
	OP_EXCL 		= '!',	
	OP_AT			= '@',
	OP_XQUOTE		= '`',
	OP_SQUOTE		= '\'',
	OP_OPEN			= '(',
	OP_CLOSE		= ')',
	OP_HASH			= '#',
	OP_BOPEN		= '[',
	OP_BBOPEN		= 136,	// "[["
	OP_BCLOSE		= ']',
	OP_BBCLOSE		= 137,	// "]]"
	OP_STAR			= '*',
	OP_COMMA		= ',',
	OP_PLUS			= '+',	// binary as usual, as unary modified label scope
	OP_MINUS		= '-',
	OP_DIV			= '/',
	OP_COLON		= ':',
	OP_SEMICOLON		= ';',
	OP_DOT			= '.',
	OP_LARGER		= '>',
	OP_LARGEROREQUAL	= 128,	// ">="
	OP_LESS			= '<',
	OP_LESSOREQUAL		= 129,	// "<="
	OP_ASSIGN		= '=',
	OP_ASSIGNADD		= 138,	// "+="
	OP_ASSIGNSUB		= 139,	// "-="
	OP_ASSIGNMULT		= 140,	// "*="
	OP_ASSIGNDIV		= 141,	// "/="
	OP_ASSIGNMOD		= 142,	// "%="
	OP_ASSIGNBITOR		= 143,	// "|="
	OP_ASSIGNBITAND		= 144,	// "&="
	OP_ASSIGNBITXOR		= 145,	// "^="
	OP_ASSIGNSHIFTLEFT	= 146,	// "<<="
	OP_ASSIGNSHIFTRIGHT	= 147,	// ">>="
	OP_BITXOR		= '^',
	OP_BITOR		= '|',
	OP_BITAND		= '&',	// binary as usual, as unary modifies label scope
	OP_EQUAL		= 138,	// "=="
	OP_NOTEQUAL		= 130,	// "!=", "<>", "><"
	OP_SHIFTLEFT		= 131,	// "<<"
	OP_SHIFTRIGHT		= 132,	// ">>"
	OP_LOGICAND		= 133,	// "&&"
	OP_LOGICOR		= 134,	// "||"
	OP_LOGICXOR		= 135,	// "^^"
	OP_BITINV		= '~',
	OP_MOD			= '%',
	// register indexing
	OP_XIND			= 148, 	// ",x"
	OP_YIND			= 149,	// ",y"
	OP_ZIND			= 150,	// ",z"
	OP_SIND			= 151,	// ",s"
	OP_BIND			= 152, 	// ",b"
} op_t;

extern int prio_of_operator_table[];

static inline int prio_of_operator(const op_t op) {
	return prio_of_operator_table[op];
}

typedef struct {
	const char 	*line;		// line to tokenize
	tok_t		type;		// type of token
	int 		ptr;		// current pointer to token in line
	int		len;		// length of current token (total, i.e. incl. quotes for a string)
	int		strptr;		// pointer to start of string for T_STRING_LITERAL
	int		strlen;		// length of string for T_STRING_LITERAL
	long		value;		// value for decimal, octal, binary, or hex literal
} tokenizer_t;

// initialize a tokenizer 
tokenizer_t *tokenizer_init(const char *line);

// set to next token; return true when there is a valid token
// if allow_index is true, comma is folded into ",x" etc addressig mode tokens
bool_t tokenizer_next(tokenizer_t *tok, int allow_index);

void tokenizer_free(tokenizer_t *tok);

#endif


