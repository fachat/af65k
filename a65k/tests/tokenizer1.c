
#include <stdbool.h>

#include "log.h"
#include "tokenizer.h"


const char *teststrings[] = {
	"label1 lda #$10",
	"label2: lda <xyz ; comment",
	"dec 0125",
	"dec 0128",
	"inc 0x12df",
	"inc $1234f",
	".byt 'xsd&'",
	".asc \"dsdd'",
	".asc \"dsAâ‚¬\"",
	NULL
};


void do_test(const char *teststr);

int main(int argc, char *argv[]) {


	printf("tokenizer:\n");

	for (int p = 0; ;p++) {
		const char *teststr = teststrings[p];
		if (teststr == NULL) {
			break;
		}

		do_test(teststr);
	}
	return 0;
}


void do_test(const char *teststr) {

	printf("tokenizing line: %s\n", teststr);

	tokenizer_t *tok = tokenizer_init(teststr);

	while (tokenizer_next(tok)) {

		printf("TOK -> type=%d, len=%d, val=%ld, ptr=%s\n", tok->type, tok->len, tok->value, tok->line+tok->ptr);
	}
	
	printf("END -> type=%d, value=%ld\n", tok->type, tok->value);

	tokenizer_free(tok);
}

