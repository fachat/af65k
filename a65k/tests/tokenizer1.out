tokenizer:
tokenizing line: label1 lda #$10
TOK -> type=2, len=6, ptr=label1 lda #$10
TOK -> type=2, len=3, ptr=lda #$10
TOK -> type=3, len=1, ptr=#$10
TOK -> type=5, len=2, ptr=10
END -> type=7
tokenizing line: label2: lda <xyz ; comment
TOK -> type=2, len=6, ptr=label2: lda <xyz ; comment
TOK -> type=3, len=1, ptr=: lda <xyz ; comment
TOK -> type=2, len=3, ptr=lda <xyz ; comment
TOK -> type=3, len=1, ptr=<xyz ; comment
TOK -> type=2, len=3, ptr=xyz ; comment
TOK -> type=3, len=1, ptr=; comment
TOK -> type=2, len=7, ptr=comment
END -> type=7
tokenizing line: dec 0125
TOK -> type=2, len=3, ptr=dec 0125
TOK -> type=5, len=4, ptr=0125
END -> type=7
tokenizing line: dec 0128
TOK -> type=2, len=3, ptr=dec 0128
END -> type=6
tokenizing line: inc 0x12df
TOK -> type=2, len=3, ptr=inc 0x12df
TOK -> type=5, len=4, ptr=12df
END -> type=7
tokenizing line: inc $1234f
TOK -> type=2, len=3, ptr=inc $1234f
TOK -> type=5, len=5, ptr=1234f
END -> type=7
tokenizing line: .byt 'xsd&'
TOK -> type=3, len=1, ptr=.byt 'xsd&'
TOK -> type=2, len=3, ptr=byt 'xsd&'
TOK -> type=4, len=6, ptr='xsd&'
END -> type=7
tokenizing line: .asc "dsdd'
TOK -> type=3, len=1, ptr=.asc "dsdd'
TOK -> type=2, len=3, ptr=asc "dsdd'
TOK -> type=4, len=6, ptr="dsdd'
END -> type=7
tokenizing line: .asc "dsA€"
TOK -> type=3, len=1, ptr=.asc "dsA€"
TOK -> type=2, len=3, ptr=asc "dsA€"
TOK -> type=4, len=8, ptr="dsA€"
END -> type=7
